File: day020/121. 买卖股票的最佳时机.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # 状态转移方程dp[i] = prices[i] - min(prices[:i])
        # 时间复杂度O(N)，空间复杂度O(1)
        if len(prices) <= 1:
            return 0
        min_price = prices[0]
        max_profit = 0
        for i in range(1, len(prices)):
            cur_profit = prices[i] - min_price
            min_price = min(min_price, prices[i])
            max_profit = max(cur_profit, max_profit)
        return max_profit            

File: day020/121. 买卖股票的最佳时机.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 1) return 0;
        int maxProfit = 0, curProfit = 0, minPrice = prices[0];
        for (int i=1; i<prices.size(); i++)
        {
            if (prices[i] < minPrice)
            {
                minPrice = prices[i];
                continue;
            }
            curProfit = prices[i] - minPrice;
            maxProfit = maxProfit > curProfit ? maxProfit : curProfit;
        }
        return maxProfit;
    }
};
